include/swift/AST/TypeRepr.h

```diff
@@ -281,165 +284,136 @@ class IdentTypeRepr;
+                                 DeclNameLoc NameLoc, DeclNameRef Name);
+
+  static DeclRefTypeRepr *create(const ASTContext &C, TypeRepr *Base,
+                                 DeclNameLoc NameLoc, DeclNameRef Name,
+                                 ArrayRef<TypeRepr *> GenericArgs,
+                                 SourceRange AngleBrackets);
+
+  /// Returns the qualifier or base type representation. For example, `A.B`
+  /// for `A.B.C`. The base of a `UnqualifiedIdentTypeRepr` is null.
+  TypeRepr *getBase() const;
+
+  /// Returns the root qualifier. For example, `A` for `A.B.C`. The root
+  /// qualifier of a `UnqualifiedIdentTypeRepr` is itself.
+  TypeRepr *getRoot();
+
+  /// Returns the root qualifier. For example, `A` for `A.B.C`. The root
+  /// qualifier of a `UnqualifiedIdentTypeRepr` is itself.
+  const TypeRepr *getRoot() const;
+
+  DeclNameLoc getNameLoc() const;
   DeclNameRef getNameRef() const;
 
   /// Replace the identifier with a new identifier, e.g., due to typo
   /// correction.
-  void overwriteNameRef(DeclNameRef newId) { IdOrDecl = newId; }
+  void overwriteNameRef(DeclNameRef newId);
 
-  /// Return true if this name has been resolved to a type decl. This happens
-  /// during type resolution.
-  bool isBound() const { return IdOrDecl.is<TypeDecl *>(); }
+  /// Returns whether this instance has been bound to a type declaration. This
+  /// happens during type resolution.
+  bool isBound() const;
 
-  TypeDecl *getBoundDecl() const { return IdOrDecl.dyn_cast<TypeDecl*>(); }
+  /// Returns the type declaration this instance has been bound to.
+  TypeDecl *getBoundDecl() const;
 
-  DeclContext *getDeclContext() const {
-    assert(isBound());
-    return DC;
-  }
+  DeclContext *getDeclContext() const;
 
-  void setValue(TypeDecl *TD, DeclContext *DC) {
-    IdOrDecl = TD;
-    this->DC = DC;
-  }
+  void setValue(TypeDecl *TD, DeclContext *DC);
 
-  static bool classof(const TypeRepr *T) {
-    return T->getKind() == TypeReprKind::SimpleIdent ||
-           T->getKind() == TypeReprKind::GenericIdent;
-  }
-  static bool classof(const IdentTypeRepr *T) { return true; }
+  /// Returns whether this instance has a generic argument list. That is, either
+  /// a valid angle bracket source range, or a positive number of generic
+  /// arguments.
+  bool hasGenericArgList() const;
 
-protected:
-  void printImpl(ASTPrinter &Printer, const PrintOptions &Opts) const;
+  unsigned getNumGenericArgs() const;
 
-  SourceLoc getLocImpl() const { return Loc.getBaseNameLoc(); }
-  friend class TypeRepr;
-};
+  ArrayRef<TypeRepr *> getGenericArgs() const;
 
-/// A simple identifier type like "Int".
-class SimpleIdentTypeRepr : public IdentTypeRepr {
-public:
-  SimpleIdentTypeRepr(DeclNameLoc Loc, DeclNameRef Id)
-      : IdentTypeRepr(TypeReprKind::SimpleIdent, Loc, Id) {}
+protected:
+  /// Returns whether this instance has a valid angle bracket source range.
+  bool hasAngleBrackets() const;
```

Comment:
Is this different from hasGenericArgsList()?

Reply:
Yeah, `hasGenericArgsList` is `hasAngleBrackets() || getNumGenericArgs() != 0`, in anticipation of an empty generic argument list or an invalid source range (unit tests).